File: netapp-manageability-sdk-9.3/src/sample/Data_ONTAP/Java/nfs/nfs.java
/*
 * $Id: //depot/prod/zephyr/belair/src/sample/java/nfs/nfs.java#1 $
 *
 * nfs.java
 * Copyright (c) 2001-2007 Network Appliance, Inc.
 * All rights reserved.
 *
 * Sample for usage of following nfs group API:
 *			nfs-enable
 *			nfs-disable
 *			nfs-status
 * 			nfs-exportfs-list-rules
 *
 * Usage:
 * nfs <filer> <user> <password> <operation> <value1> [<value2>]
 *
 * <filer>      -- Name/IP address of the filer
 * <user>       -- User name
 * <password>   -- Password
 * <operation>  --
 *		enable - Enable nfs
 *		disable - Disable nfs
 *		status - nfs status
 * 		list   - list nfs export rules
 *
 */

import java.io.IOException;
import java.util.*;
import netapp.manage.*;

public class nfs {

    public static void print_usage() {
        System.out
                .println("\nUsage : nfs <filer> <user> <passwd> <operation> \n");
        System.out.print("<filer>	-- ");
        System.out.println("Name/IP address of the filer");
        System.out.println("<user>	-- User Name");
        System.out.println("<passwd>	-- Password");
        System.out.println("<operation>	--");
        System.out.println("\tenable - To enable NFS Service");
        System.out.println("\tdisable - To disable NFS Service");
        System.out.println("\tstatus - To print the status of NFS Service");
        System.out.println("\tlist - To list the NFS export rules");
        System.exit(1);
    }

    public static void main(String[] args) {
        NaElement xi;
        NaElement xo;
        NaServer s;
        String operation;
        NaElement pathnames;
        int path_counter;
        NaElement xc;

        if (args.length < 4) {
            print_usage();
        }
        try {
            /*
             * Initialize connection to server, and request version 1.3 of the
             * API set
             */
            s = new NaServer(args[0], 1, 3);

            /* Set connection style(HTTP) */
            s.setStyle(NaServer.STYLE_LOGIN_PASSWORD);
            s.setAdminUser(args[1], args[2]);
            operation = args[3];

            /*
             * To invoke nfs-enable API Usage: nfs <filer> <user> <password>
             * enable
             */
            if (operation.equals("enable")) {

                xi = new NaElement("nfs-enable");
                xo = s.invokeElem(xi);
                System.out.println("enabled successfully!");
            }

            /*
             * To invoke nfs-enable API Usage: nfs <filer> <user> <password>
             * disable
             */
            else if (operation.equals("disable")) {

                xi = new NaElement("nfs-disable");
                xo = s.invokeElem(xi);
                System.out.println("disabled successfully!");
            }

            /*
             * To invoke nfs-status API Usage: nfs <filer> <user> <password>
             * status
             */
            else if (operation.equals("status")) {

                xi = new NaElement("nfs-status");
                xo = s.invokeElem(xi);
                String enabled = xo.getChildContent("is-enabled");
                if (enabled.compareTo("true") == 0) {
                    System.out.println("NFS Server is enabled");
                } else {
                    System.out.println("NFS Server is disabled");
                }
            }

            /*
             * To invoke nfs-exportfs-list-rules API Usage: nfs <filer> <user>
             * <password> list
             */
            else if (operation.equals("list")) {

                xi = new NaElement("nfs-exportfs-list-rules");
                xo = s.invokeElem(xi);

                List retList = xo.getChildByName("rules").getChildren();
                Iterator retIter = retList.iterator();

                while (retIter.hasNext()) {

                    NaElement retInfo = (NaElement) retIter.next();
                    String path_name = retInfo.getChildContent("pathname");
                    String rw_list = new String("rw=");
                    String ro_list = new String("ro=");
                    String root_list = new String("root=");

                    if (retInfo.getChildByName("read-only") != null) {
                        NaElement rule_elem = retInfo
                                .getChildByName("read-only");
                        List hosts = rule_elem.getChildren();
                        Iterator hostIter = hosts.iterator();
                        while (hostIter.hasNext()) {

                            NaElement hostInfo = (NaElement) hostIter.next();
                            if (hostInfo.getChildContent("all-hosts") != null) {
                                String allhost = hostInfo
                                        .getChildContent("all-hosts");
                                if (allhost.compareTo("true") == 0) {
                                    ro_list = ro_list + "all-hosts";
                                    break;
                                }
                            } else if (hostInfo.getChildContent("name") != null) {
                                ro_list = ro_list
                                        + hostInfo.getChildContent("name")
                                        + ":";
                            }

                        }
                    }
                    if (retInfo.getChildByName("read-write") != null) {
                        NaElement rule_elem = retInfo
                                .getChildByName("read-write");
                        List hosts = rule_elem.getChildren();
                        Iterator hostIter = hosts.iterator();
                        while (hostIter.hasNext()) {

                            NaElement hostInfo = (NaElement) hostIter.next();
                            if (hostInfo.getChildContent("all-hosts") != null) {
                                String allhost = hostInfo
                                        .getChildContent("all-hosts");
                                if (allhost.compareTo("true") == 0) {
                                    rw_list = rw_list + "all-hosts";
                                    break;
                                }
                            } else if (hostInfo.getChildContent("name") != null) {
                                rw_list = rw_list
                                        + hostInfo.getChildContent("name")
                                        + ":";
                            }

                        }
                    }
                    if (retInfo.getChildByName("root") != null) {
                        NaElement rule_elem = retInfo.getChildByName("root");
                        List hosts = rule_elem.getChildren();
                        Iterator hostIter = hosts.iterator();
                        while (hostIter.hasNext()) {

                            NaElement hostInfo = (NaElement) hostIter.next();
                            if (hostInfo.getChildContent("all-hosts") != null) {
                                String allhost = hostInfo
                                        .getChildContent("all-hosts");
                                if (allhost.compareTo("true") == 0) {
                                    root_list = root_list + "all-hosts";
                                    break;
                                }
                            } else if (hostInfo.getChildContent("name") != null) {
                                root_list = root_list
                                        + hostInfo.getChildContent("name")
                                        + ":";
                            }

                        }
                    }
                    path_name = path_name + "  ";
                    if (ro_list.compareTo("ro=") != 0) {
                        path_name = path_name + ro_list;
                    }
                    if (rw_list.compareTo("rw=") != 0) {
                        path_name = path_name + "," + rw_list;
                    }
                    if (root_list.compareTo("root=") != 0) {
                        path_name = path_name + "," + root_list;
                    }

                    System.out.println(path_name);
                }

            } else {
                print_usage();
            }
            s.close();
        } catch (NaAuthenticationException e) {
            System.err.println("Bad login/password");
        } catch (NaAPIFailedException e) {
            System.err.println("API failed (" + e.getReason() + ")");
        } catch (IOException e) {
            e.printStackTrace();
        } catch (NaProtocolException e) {
            e.printStackTrace();
        } catch (Exception e) {
            e.printStackTrace();
        }
    }
}

File: netapp-manageability-sdk-9.3/src/sample/Data_ONTAP/DotNet/CSharp/nfs/nfs.cs
//======================================================================//
//                                                                      //
// $ID$                                                                 //
//                                                                      //
// nfs.cs                                                               //
//                                                                      //
// Brief information of the contents                                    //
//                                                                      //
// Copyright 2008 NetApp. All rights reserved. Specifications subject   //
// to change without notice.                                            //
//                                                                      //
// This SDK sample code is provided AS IS, with no support or           //
// warranties of any kind, including but not limited to                 //
// warranties of merchantability or fitness of any kind,                //
// expressed or implied.  This code is subject to the license           //
// agreement that accompanies the SDK.                                  //
//                                                                      //
// Sample for usage of following nfs group API:                         //
//                      nfs-enable                                      //
//                      nfs-disable                                     //
//                      nfs-status                                      //
//                      nfs-exportfs-list-rules                         //
//                                                                      //
// Usage:                                                               //
// nfs <filer> <user> <password> <operation>                            //
//                                                                      //
// <filer>      -- Name/IP address of the filer                         //
// <user>       -- User name                                            //
// <password>   -- Password                                             //
// <operation>  --                                                      //
//                 enable                                               //
//                 disable                                              //
//                 status                                               //
//                 list                                                 //
//======================================================================//

using System; 
using System.Collections.Generic;
using System.Text;
using NetApp.Manage;

namespace nfs
{
    class Nfs
    {
        public static void PrintUsage()
        {
            Console.Out.WriteLine("\nUsage : nfs <filer> <user> <passwd> <operation> \n");
            Console.Out.Write("<filer>	-- ");
            Console.Out.WriteLine("Name/IP address of the filer");
            Console.Out.WriteLine("<user>	-- User Name");
            Console.Out.WriteLine("<passwd>	-- Password");
            Console.Out.WriteLine("<operation>	--");
            Console.Out.WriteLine("\tenable - To enable NFS Service");
            Console.Out.WriteLine("\tdisable - To disable NFS Service");
            Console.Out.WriteLine("\tstatus - To print the status of NFS Service");
            Console.Out.WriteLine("\tlist - To list the NFS export rules");
            Environment.Exit(1);
        }

        static void Main(string[] args)
        {
            NaElement xi;
		    NaElement xo;
		    NaServer s;
		    String operation;
		        		
		    if (args.Length < 4)
		    {
			    PrintUsage();
		    }
		    try
		    {
                Console.WriteLine("|-----------------------------------------|");
                Console.WriteLine("| Program to demo use of NFS related APIs |");
                Console.WriteLine("|-----------------------------------------|\n");
                /*
			    * Initialize connection to server, and
			    * request version 1.3 of the API set
			    */
			    s = new NaServer(args[0], 1, 3);
    			
			    /* Set connection style(HTTP)*/
			    s.Style = NaServer.AUTH_STYLE.LOGIN_PASSWORD;
			    s.SetAdminUser(args[1], args[2]);
			    operation = args[3];
    			
			    /* To invoke nfs-enable API
			    *  Usage: nfs <filer> <user> <password> enable
			    *
			    */
			    if (operation.Equals("enable"))
			    {
    				
				    xi = new NaElement("nfs-enable");
				    xo = s.InvokeElem(xi);
				    Console.Out.WriteLine("enabled successfully!");
			    }
			    /* To invoke nfs-enable API
			    *  Usage: nfs <filer> <user> <password> disable
			    *
			    */
			    else if (operation.Equals("disable"))
			    {
    				
				    xi = new NaElement("nfs-disable");
				    xo = s.InvokeElem(xi);
				    Console.Out.WriteLine("disabled successfully!");
			    }
			    /* To invoke nfs-status API
			    *  Usage: nfs <filer> <user> <password> status
			    *
			    */
			    else if (operation.Equals("status"))
			    {
    				
				    xi = new NaElement("nfs-status");
				    xo = s.InvokeElem(xi);
				    String enabled = xo.GetChildContent("is-enabled");
				    if (String.Compare(enabled, "true") == 0)
				    {
					    Console.Out.WriteLine("NFS Server is enabled");
				    }
				    else
				    {
					    Console.Out.WriteLine("NFS Server is disabled");
				    }
			    }
			    /* To invoke nfs-exportfs-list-rules API
			    *  Usage: nfs <filer> <user> <password> list
			    *
			    */
			    else if (operation.Equals("list"))
			    {
    				
				    xi = new NaElement("nfs-exportfs-list-rules");
				    xo = s.InvokeElem(xi);
    				
				    System.Collections.IList retList = xo.GetChildByName("rules").GetChildren();
                    System.Console.WriteLine(xo.ToPrettyString(""));
                    Environment.Exit(0);
				    System.Collections.IEnumerator retIter = retList.GetEnumerator();
    				
				    while (retIter.MoveNext())
				    {
    					
					    NaElement retInfo = (NaElement) retIter.Current;
					    String pathName = retInfo.GetChildContent("pathname");
					    String rwList = "rw=";
					    String roList = "ro=";
					    String rootList = "root=";
    					
    					
					    if (retInfo.GetChildByName("read-only") != null)
					    {
						    NaElement ruleElem = retInfo.GetChildByName("read-only");
						    System.Collections.IList hosts = ruleElem.GetChildren();
						    System.Collections.IEnumerator hostIter = hosts.GetEnumerator();
						    while (hostIter.MoveNext())
						    {
    							
							    NaElement hostInfo = (NaElement) hostIter.Current;
							    if (hostInfo.GetChildContent("all-hosts") != null)
							    {
								    String allHost = hostInfo.GetChildContent("all-hosts");
								    if (String.Compare(allHost, "true") == 0)
								    {
									    roList = roList + "all-hosts";
									    break;
								    }
							    }
							    else if (hostInfo.GetChildContent("name") != null)
							    {
								    roList = roList + hostInfo.GetChildContent("name") + ":";
							    }
						    }
					    }
					    if (retInfo.GetChildByName("read-write") != null)
					    {
						    NaElement ruleElem = retInfo.GetChildByName("read-write");
						    System.Collections.IList hosts = ruleElem.GetChildren();
						    System.Collections.IEnumerator hostIter = hosts.GetEnumerator();
						    while (hostIter.MoveNext())
						    {
    							
							    NaElement hostInfo = (NaElement) hostIter.Current;
							    if (hostInfo.GetChildContent("all-hosts") != null)
							    {
								    String allHost = hostInfo.GetChildContent("all-hosts");
								    if (String.Compare(allHost, "true") == 0)
								    {
									    rwList = rwList + "all-hosts";
									    break;
								    }
							    }
							    else if (hostInfo.GetChildContent("name") != null)
							    {
								    rwList = rwList + hostInfo.GetChildContent("name") + ":";
							    }
						    }
					    }
					    if (retInfo.GetChildByName("root") != null)
					    {
						    NaElement ruleElem = retInfo.GetChildByName("root");
						    System.Collections.IList hosts = ruleElem.GetChildren();
						    System.Collections.IEnumerator hostIter = hosts.GetEnumerator();
						    while (hostIter.MoveNext())
						    {
    							
							    NaElement hostInfo = (NaElement) hostIter.Current;
							    if (hostInfo.GetChildContent("all-hosts") != null)
							    {
								    String allHost = hostInfo.GetChildContent("all-hosts");
								    if (String.Compare(allHost, "true") == 0)
								    {
									    rootList = rootList + "all-hosts";
									    break;
								    }
							    }
							    else if (hostInfo.GetChildContent("name") != null)
							    {
								    rootList = rootList + hostInfo.GetChildContent("name") + ":";
							    }
						    }
					    }

					    if (String.Compare(roList, "ro=") != 0)
					    {
						    pathName = pathName + ", \t" + roList;
					    }
					    if (String.Compare(rwList, "rw=") != 0)
					    {
                            pathName = pathName + ", \t" + rwList;
					    }
					    if (String.Compare(rootList, "root=") != 0)
					    {
                            pathName = pathName + ", \t" + rootList;
					    }
    					
					    Console.Out.WriteLine(pathName);
				    }
			    }
			    else
			    {
				    PrintUsage();
			    }
		    }
		    catch (NaAuthException e)
		    {
			    Console.Error.WriteLine("Bad login/password" + e.Message);
		    }
		    catch (NaApiFailedException e)
		    {
			    Console.Error.WriteLine("API failed (" + e.Message + ")");
		    }
	        catch (NaProtocolException e)
		    {
			    Console.Error.WriteLine(e.StackTrace);
		    }
		    catch (System.Exception e)
		    {
			   Console.Error.WriteLine(e.Message);
		    }
        }
    }
}
